hydrolib/core/basemodel.py

```diff
@@ -38,22 +75,45 @@ class FileModel(BaseModel, ABC):
 
     filepath: Optional[Path] = None
 
-    def __init__(self, filepath: Optional[Path] = None, *args, **kwargs) -> None:
+    def __init__(self, filepath: Optional[Path] = None, *args, **kwargs):
         """Initialize a model.
 
         The model is empty (with defaults) if no `filepath` is given,
         otherwise the file at `filepath` will be parsed."""
         # Parse the file if path is given
+        token = None
         if filepath:
-            data = self._load(Path(filepath))  # so we also accept strings
+            filepath = Path(filepath)  # so we also accept strings
+
+            # If a context is set, use it
+            folder = context_dir.get(None)
+            if folder and not filepath.is_absolute():
+                logger.info(f"Used context to get {folder} for {filepath}")
+                filepath = folder / filepath
+            # Otherwise we're the root filepath
+            # and should set the context
+            else:
+                logger.info(f"Set context to {filepath.parent}")
+                token = context_dir.set(filepath.parent)
+
+            data = self._load(filepath)
             data["filepath"] = filepath
             kwargs.update(data)
         super().__init__(*args, **kwargs)
 
+        # Once the model has been completely initialized
+        # reset the context
+        if token:
+            logger.info(f"Reset context.")
+            context_dir.reset(token)
```

Comment:
If I see a comment line above a bit of code, I'd be inclined to replace those with a separate private function which incorporates the comment. This way your code is self documenting, and it is harder for the code to go out of sync with it's description, e.g.:

```suggestion
        self._reset_context_dir(token)
```
and then add
```python
    def _reset_context_dir(token)
        if token:
            logger.info(f"Reset context.")
            context_dir.reset(token)
```